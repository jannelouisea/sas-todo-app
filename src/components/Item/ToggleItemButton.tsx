import IconButton from '@mui/joy/IconButton';
import DoneIcon from '@mui/icons-material/Done';
import PropTypes from 'prop-types';

import {
    MUIColor,
    MUIVariant,
    MUISize
} from 'src/static/enums'

type Props = {
    color: MUIColor,
    variant: MUIVariant,
    onClick: React.MouseEventHandler<HTMLButtonElement>,
}

function ToggleItemButton({ color, variant, onClick }: Props) {
    return (
        <IconButton
            color={color}
            id='complete-item-button'
            variant={variant}
            aria-label='Complete item button'
            onClick={onClick}
            sx={{ borderRadius: '50%' }}
            size={MUISize.Small}
        >
            <DoneIcon fontSize='inherit' />
        </IconButton>
    );
}

ToggleItemButton.propTypes = {
    color: PropTypes.oneOf([...Object.values(MUIColor)]),
    variant: PropTypes.oneOf([...Object.values(MUIVariant)]),
    onClick: () => { }
}

ToggleItemButton.defaultProps = {
    color: MUIColor.Primary,
    variant: MUIVariant.Outlined,
    onClick: () => { }
}

export default ToggleItemButton;