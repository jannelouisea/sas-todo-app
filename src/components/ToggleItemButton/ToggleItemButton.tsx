import IconButton from '@mui/joy/IconButton';
import DoneIcon from '@mui/icons-material/Done';
import PropTypes from 'prop-types';

import { todoItemToggled, todoItemsSorted } from 'src/features/todoItems/todoItemsSlice'
import { useAppDispatch, useAppSelector } from 'src/app/hooks';

import {
    MUIColor,
    MUIVariant,
    MUISize
} from 'src/static/enums'

type Props = {
    itemId: string,
    color: MUIColor,
    variant: MUIVariant,
}

function ToggleItemButton({ itemId, color, variant }: Props) {
    const dispatch = useAppDispatch();
    const sortBy = useAppSelector(state => state.todoItemsFilters.sortBy);

    const toggleItem = () => {
        dispatch(todoItemToggled(itemId));
        dispatch(todoItemsSorted(sortBy));
    }

    return (
        <IconButton
            color={color}
            id='complete-item-button'
            variant={variant}
            aria-label='Complete item button'
            onClick={toggleItem}
            sx={{ borderRadius: '50%' }}
            size={MUISize.Small}
        >
            <DoneIcon fontSize='inherit' />
        </IconButton>
    );
}

ToggleItemButton.propTypes = {
    itemId: PropTypes.string,
    color: PropTypes.oneOf([...Object.values(MUIColor)]),
    variant: PropTypes.oneOf([...Object.values(MUIVariant)]),
}

ToggleItemButton.defaultProps = {
    color: MUIColor.Primary,
    variant: MUIVariant.Outlined,
}

export default ToggleItemButton;