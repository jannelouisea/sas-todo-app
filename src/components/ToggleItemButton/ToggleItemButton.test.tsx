import renderer from 'react-test-renderer';
import { render, screen, fireEvent } from '@testing-library/react';
import ToggleItemButton from './ToggleItemButton';

import {
    MUIColor,
    MUIVariant,
} from 'src/static/enums'

describe('ToggleItemButton', () => {
    it('renders correctly', () => {
        const tree = renderer
            .create(<ToggleItemButton />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('calls onClick prop function on click', () => {
        const onClickSpy = jest.fn();
        render(
            <ToggleItemButton
                color={MUIColor.Success}
                variant={MUIVariant.Soft}
                onClick={onClickSpy}
            />
        );
        fireEvent.click(screen.getByLabelText('Complete item button'));
        expect(onClickSpy).toHaveBeenCalledTimes(1);
    })
});
