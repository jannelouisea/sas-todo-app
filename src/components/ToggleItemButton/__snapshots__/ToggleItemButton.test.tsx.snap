// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleItemButton renders correctly 1`] = `
<button
  aria-label="Complete item button"
  className="MuiIconButton-root MuiIconButton-variantOutlined MuiIconButton-colorPrimary MuiIconButton-sizeSm css-1du9397-JoyIconButton-root"
  id="complete-item-button"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
>
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
    data-testid="DoneIcon"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M9 16.2 4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
    />
  </svg>
</button>
`;
